rust/kernel/sync/arc.rs

```diff
@@ -475,6 +512,44 @@ impl<T> UniqueArc<MaybeUninit<T>> {
     /// Converts a `UniqueArc<MaybeUninit<T>>` into a `UniqueArc<T>` by writing a value into it.
     pub fn write(mut self, value: T) -> UniqueArc<T> {
         self.deref_mut().write(value);
+        // SAFETY: we just initialized `self`
+        unsafe { self.assume_init() }
+    }
+
+    /// Initialize the value after creating the `UniqueArc`.
+    #[allow(clippy::type_complexity)]
+    pub fn pin_init_now<E>(
+        mut self,
+        init: impl PinInit<T, E>,
+    ) -> core::result::Result<Pin<UniqueArc<T>>, (E, UniqueArc<MaybeUninit<T>>)> {
+        unsafe {
+            match init.__pinned_init(self.deref_mut().as_mut_ptr()) {
+                Ok(()) => Ok(Pin::from(self.assume_init())),
+                Err(e) => Err((e, self)),
+            }
+        }
+    }
+
+    /// Initialize the value after creating the `UniqueArc`.
+    pub fn init_now<E>(
```

Comment:
Is this really necessary? I suppose we can just have a pinned variant, and if someone needs an unpinned version then they can take it out from `Pin` themself.

Note that our `Arc` is different from Rust std `Arc`, it is pinned by default. Our `UniqueArc` isn't pinned by default, but we probably could. @wedsonaf do you think it make sense to have `UniqueArc` pinned and only provide a `DerefMut` if `T: Unpin`? We can add an `as_mut` function that gives `Pin<&mut Self>` for `T: !Unpin`.

Reply:
I added it for completeness. We could surely just remove it. However it might occur that someone has some data that is `!Unpin` and they want to still be able to mutate it before pinning... (no idea how likely that is though, probably very unlikely)